### PR TITLE
DCGAN model

### PR DESCRIPTION
**Pull request by harbac**
This PR introduces DCGAN architectures into deeplay. The models are implemented as described in the original DCGAN article ([link](https://arxiv.org/abs/1511.06434)).

**Example:**

Creating DCGAN generator
```
generator = dl.DcganGenerator(
    latent_dim=100,
    features_dim=64,
    output_channels=1,
    class_conditioned_model=False,
    num_classes=10, # only used if class_conditioned_model=True
    embedding_dim=100, # only used if class_conditioned_model=True
)
```

Creating DCGAN discriminator
```
discriminator = dl.DcganDiscriminator(
    input_channels=1,
    features_dim=64,
    class_conditioned_model=False,
    num_classes=10,  # only used if class_conditioned_model=True
    embedding_dim=100,  # only used if class_conditioned_model=True
)
```

As DCGAN is a fixed architecture, the images need to be resized to `64 x 64` before passing through the models. The number of Conv blocks in discriminator and the ConvTranspose blocks in the generator are fixed to 4 as per the paper. However, the number of features in each block can be controlled by changing the `features_dim` attribute.

For generator, number of features in each ConvTranspose block goes as `[features_dim*16, features_dim*8, features_dim*4, features_dim*2]`

For discriminator, the number of features in each Conv block goes by `[features_dim, features_dim*2, features_dim*4, features_dim*8]`.

By changing the `features_dim` parameter, the relative strength of generator and discriminator can be controlled.

### CODE REVIEW

**Review by giovannivolpe — APPROVED** [review]
@BenjaminMidtvedt please check carefully

File: deeplay/components/dcgan/dcgan.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+                if batch_norm
+                else Layer(nn.Identity),
+                Layer(nn.LeakyReLU, 0.2),
+            )
+        )
+
+    def forward(self, x):
+        for block in self.blocks:
+            x = block(x)
+        return x
```
Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: Should be defined in components/blocks, and subclass `LayerActivationNormalization`. Something like

```py
class ConvBlock(LayerActivationNormalization):
    def __init__(
        self,
        in_channels,
        out_channels,
        kernel_size,
        stride,
        padding,
        activation=None,
        normalization=None,
    ):
    super().__init__(
         layer=Layer(Conv2d, kernel_size=kernel_size, ...),
         activation=activation or Layer(ReLU),
         normalization=normalization or Layer(BatchNorm2d, out_channels)
    )

File: deeplay/components/dcgan/dcgan.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+    """
+
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        kernel_size,
+        stride,
+        padding,
+        batch_norm=True,
```
Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: argument should be `normalization` which accepts a deeplaymodule. Should also add argument for activation

File: deeplay/components/dcgan/dcgan.py [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+            )
+        )
+
+    def forward(self, x):
+        for block in self.blocks:
+            x = block(x)
+        return x
+
+
+class DcganDiscriminator(DeeplayModule):
```
Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: I prefer DCGANDiscriminator.

Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: Since it's a exact architecture, we should move to something like models/dcgan/discriminator.py

We should also decide if we use the suggestion by Jesus to make it a function

File: deeplay/components/dcgan/dcgan.py [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+        activation: Optional[Type[nn.Module]] = None,
+        normalization: Optional[Type[nn.Module]] = None,
+        **kwargs: Any,
+    ) -> None:
+        ...
+
+    configure = DeeplayModule.configure
+
+
+class ConvTransposeBlock(DeeplayModule):
```
Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: Same comments as for ConvBlock.

File: deeplay/components/dcgan/dcgan.py [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+            )
+        )
+
+    def forward(self, x):
+        for block in self.blocks:
+            x = block(x)
+        return x
+
+
+class DcganGenerator(DeeplayModule):
```
Comment (BenjaminMidtvedt) [inline deeplay/components/dcgan/dcgan.py]: Same comments as for discriminator

**Review by BenjaminMidtvedt — COMMENTED** [review]
Mostly organization-based comments. The blocks specifically should have a different inheritance

File: deeplay/components/dcgan/dcgan.py (hunk 1 above)
Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Fixed this now. However, it is still at dcgan/dcgan.py. Do you want me to move it somewhere else?

File: deeplay/components/dcgan/dcgan.py (hunk 2 above)
Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Corrected.

File: deeplay/components/dcgan/dcgan.py (hunk 3 above)
Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Changed.

Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Yeah, I'm holding on the file relocations until we decide on it. 

For the second comment, for fixed architectures I'm not sure if it is useful to make it as function. Since the blocks would be specific to the fixed architectures, it would be good to leave them standalone if people want to check how it is built.

File: deeplay/components/dcgan/dcgan.py (hunk 4 above)
Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Corrected.

File: deeplay/components/dcgan/dcgan.py (hunk 5 above)
Comment (harbac) [reply deeplay/components/dcgan/dcgan.py]: Corrected.

**Comment by BenjaminMidtvedt** [comment]
I think it's looking good now, nice work! I'm good to merge

**Review by BenjaminMidtvedt — APPROVED** [review]
